### ISSUE TITLE
Data logging

### ISSUE THREAD

**Issue by milenovic** [thomasnordquist/MQTT-Explorer#632]
Is there an option to save the data in any way? In another words, to use it as a data logger.

**Comment by hanssan** [comment]
Congrats to this fine tool!
I support milenovic' comment: A data logger would be very useful to analze interactions via mqtt (history is too cumbersome for that). Ideally a dump of the sequence of received messages to a a file would serve the purpose.

**Comment by guillochon** [comment]
I third this request, I'd love to be able to export data using this tool to CSV files rather than having to use a separate tool to do so.

**Comment by squintz** [comment]
Another request. Data logging would be very useful in tracking down issues during development.

**Comment by nicklasb** [comment]
Yes, logging whatever histories that are created into a file would be awesome. 
(also because it seems to lose the history sometime)
Great tool!

**Comment by ZyroxCo** [comment]
where is the folder that data is logged in?

**Comment by rebatoma** [comment]
Agree, it would be nice to have the possibility to extract a log file with the MQTT messages received on the broker, while the MQTT-Explorer client was running.

**Comment by YSaxon** [comment]
Open DevTools (should be an option in the file menu)
Now in the JS Console
```
const reactRoot = document.querySelector('#app')._reactRootContainer;
const store = reactRoot._internalRoot.current.child.memoizedProps.store;
var tree = store.getState().connection.tree
```

From `tree` you can look at `tree.cachedChildTopics` or `tree.messageHistory.items`, or if you are paused, you can use `tree.unmergedMessages.buffer`. And then you can parse that out into a csv if you want.

Here's an example:
```
var cachedChildTopics = tree.cachedChildTopics;
var rows = [];

for (var i = 0; i < cachedChildTopics.length; i++) {
    var treeNode = cachedChildTopics[i];
    var topic = treeNode.mqttMessage.topic;
    var messageHistory = treeNode.messageHistory.items;

    for (var j = 0; j < messageHistory.length; j++) {
        var item = messageHistory[j];
        var timestamp = item.received;
        var base64Message = item.value.base64Message;
        var rawMessage = base64Decode(base64Message);
        var hexMessage = rawToHex(rawMessage);

        rows.push({
            topic: topic,
            clientId: clientId,
            timestamp: timestamp,
            base64Message: base64Message,
            rawMessage: rawMessage,
            hexMessage: hexMessage
        });
    }
}

var csvContent = "topic,timestamp,base64Message,rawMessage,hexMessage\n";
for (var i = 0; i < rows.length; i++) {
    var row = rows[i];
    csvContent += [row.topic, row.timestamp, row.base64Message, row.rawMessage, 
                   row.hexMessage].join(",") + "\n";
}

console.log(csvContent);

```

**Comment by frankvandenhurk** [comment]
Thanks for the code, I'd love to get that workling. Unfortunately, I get:

VM378:44 Uncaught TypeError: Cannot read property 'base64Message' of undefined
    at window.traverseTree (<anonymous>:44:44)
    at window.traverseTree (<anonymous>:35:17)
    at window.traverseTree (<anonymous>:35:17)
    at window.traverseTree (<anonymous>:35:17)
    at window.traverseTree (<anonymous>:35:17)
    at window.traverseTree (<anonymous>:35:17)
    at <anonymous>:59:1

What am I doing wrong?

(I opend the dev tools, clicked on the Console tab, copied the first snipped (3 lines), it returned "undefined" and than I copied the long snippet, which resulted in the error)

**Comment by Stephen-Degould** [comment]
Hello Frank,

If you're still struggling you simply need to update 
`var base64Message = item.value.base64Message `
to
`var base64Message = item.payload.base64Message`
Then it all seems to work okay.

**Comment by jonathandavidhopper** [comment]
I get the same 

> Thanks for the code, I'd love to get that workling. Unfortunately, I get:
> 
> VM378:44 Uncaught TypeError: Cannot read property 'base64Message' of undefined at window.traverseTree (:44:44) at window.traverseTree (:35:17) at window.traverseTree (:35:17) at window.traverseTree (:35:17) at window.traverseTree (:35:17) at window.traverseTree (:35:17) at :59:1
> 
> What am I doing wrong?
> 
> (I opend the dev tools, clicked on the Console tab, copied the first snipped (3 lines), it returned "undefined" and than I copied the long snippet, which resulted in the error)

I get the same error, and the suggested fix of 
 update
var base64Message = item.value.base64Message 
to
var base64Message = item.payload.base64Message
makes no difference...

Any suggestions?

**Comment by Henning-Lenz** [comment]
I came here because this is one of the most wanted features I'd like to see in this software to easily export some limited time-series. Currently working with version 0.4.0-beta.6.

I used the code of @YSaxon and had to apply the change of @Stephen-Degould (many thanks for the template) and then tried to make my way through (not a programmer by training):
- There was some data that mosquitto dumped below the "$SYS" topic that the code was not able to cope with, therefore I started where interesting data is by calling the function with some other node like `traverseTree(tree.edges['shellies'].target, "/shellies");`. With that it successfully returned data as CSV.
- After MQTT Explorer ran for some time the code stopped working because early messages get removed and therefore the message history array does not contain indexes starting at 0. That may be what @jonathandavidhopper also struggled with. Therefore I looped elements as given in other part of the code.
- While exploring the tree I found `item.payload.unicodeValue` which seems to contain the already decoded data which (at least in my case) is enough to export.
- Additionally applying `toISOString()` to the timestamp or even adding `slice(0,-1)` for removing the "Z" for output makes data more easy to parse depending on which software you are using downstream (e.g. LibreOffice does not like the "Z" for automatic parsing).

Rearranged some functions/global vars and ended with some useable result for my use case which is complete, but stripped down compared to above code:

```
const reactRoot = document.querySelector('#app')._reactRootContainer;
const store = reactRoot._internalRoot.current.child.memoizedProps.store;
var tree = store.getState().connection.tree

window.parseHistoryElement = function(topic, index, histelement) {
    var timestamp = histelement.received.toISOString();
    // LibreOffice does not like "Z", therefore remove it
    timestamp = timestamp.slice(0,-1);

    return {
        topic: topic,
        messageIndex: index,
        messageNumber: histelement.messageNumber,
        timestamp: timestamp,
        unicodeMessage: histelement.payload.unicodeValue
    };
}

// Function to traverse the tree recursively
window.traverseTree = function(treeNode, topic) {
    var rows = [];

    if (treeNode.edges) {
        for (var edgeName in treeNode.edges) {
            if (treeNode.edges.hasOwnProperty(edgeName)) {
                var newrows = traverseTree(treeNode.edges[edgeName].target, topic + '/' + edgeName)
                rows = rows.concat(newrows);
            }
        }
    }
    if (treeNode.messageHistory && treeNode.messageHistory.items) {
        var messageHistory = treeNode.messageHistory.items;
        for (var idx in messageHistory) {
            var histparsed = parseHistoryElement(topic, idx, messageHistory[idx]);
            rows.push(histparsed);
        }
    }
    
    return rows;
}

window.writeCSV = function(data) {
    // Init CSV with header
    var csvContent = "topic,messageNumber,timestamp,unicodeMessage\n";

    for (var i = 0; i < data.length; i++) {
        var row = data[i];
        csvContent += [row.topic, row.messageIndex, row.messageNumber, row.timestamp, row.unicodeMessage].join(",") + "\n";
    }

    return csvContent;
}

var data_array = traverseTree(tree.edges['shellies'].target, "/shellies");

var data_csv = writeCSV(data_array);

console.log(data_csv);
```
